netket/utils/array.py

```diff
@@ -44,13 +54,26 @@ def __pre_init__(self, wrapped):
             if isinstance(wrapped, np.ndarray):
                 wrapped.flags.writeable = False
 
-        return (wrapped,), {}
+        self._wrapped: np.array = wrapped
+        self._hash: Optional[int] = None
+
+    @property
+    def wrapped(self):
+        """The read-only wrapped array."""
+        return self._wrapped
 
     def __hash__(self):
-        return hash(self.wrapped.tobytes())
+        if self._hash is None:
+            self._hash = hash(self.wrapped.tobytes())
+        return self._hash
 
     def __eq__(self, other):
-        return type(other) is HashableArray and np.all(self.wrapped == other.wrapped)
+        return (
+            type(other) is HashableArray
+            and self.shape == other.shape
+            and self.dtype == other.dtype
+            and np.all(self.wrapped == other.wrapped)
```

Comment:
I think we should compare hashes, otherwise jax might compare objects and that can be expensive...